plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemory.hs

```diff
@@ -129,33 +129,90 @@ instance Pretty ExCPU where
 instance PrettyBy config ExCPU where
     prettyBy _ m = pretty m
 
+{- Note [ExMemoryUsage instances for non-constants]
+In order to calculate the cost of a built-in function we need to feed the 'ExMemory' of each
+argument to the costing function associated with the builtin. For a polymorphic builtin this means
+that we need to be able to compute the 'ExMemory' of the AST provided as an argument to the builtin.
+How do we do that? Our strategy is:
+
+1. if the AST is a wrapped constant, then calculate the 'ExMemory' of the constant
+2. if the AST is something else, return 1
+
+This is pretty reasonable: a polymorphic builtin *is* allowed to check if the AST that it got as an
+argument is a constant or not, and if it happens to be a constant, the builtin *is* allowed to use
+it whatever way it wishes (see Note [Builtins and Plutus type checking] for details). Hence a
+builtin may in fact do something ad hoc for constants and we need to account for this possibility in
+the costing machinery.
+
+But if the given AST is not a constant, the builtin can't do anything else with it, hence we simply
+return 1, meaning "the costing function can't use this 'ExMemory' in any non-vacuous way".
+
+See 'HasMeaningIn' for a full list of constraints determining what a builtin can do with values.
+
+However for all types of values, except the one used by the production evaluator, we implement
```

Comment:
So "value" here means the specific notion of value used by the evaluator, not just a term that can't be reduced. That makes sense since in our evaluators (or at least the ones we have at the moment) we have a separate type of `Value`, but it took me a little time to see what you meant. I can't think of a better way to put it though.

Reply:
Yeah, the "value" term (pardon the pun) is way too overloaded.